tests/helper.py

```diff
@@ -90,18 +90,29 @@ def assert_failed(self, with_errors=False, with_warnings=None):
             fail_on_warnings=with_warnings or with_warnings is None), \
                 "Test did not fail; failure was expected."
 
-        if with_errors:
-            assert self.err.errors, "Errors were expected."
-        elif self.err.errors:
-            raise AssertionError("Tests found unexpected errors: %s" %
-                                self.err.print_summary(verbose=True))
-
-        if with_warnings is not None:
-            if with_warnings:
-                assert self.err.warnings, "Warnings were expected."
-            elif self.err.warnings:
-                raise ("Tests found unexpected warnings: %s" %
-                           self.err.print_summary())
+        def check_message(message, props):
+            return all(message.get(prop) == value
+                       for prop, value in props.iteritems())
+
+        def find_message(messages, props):
+            return any(check_message(message, props)
+                       for message in messages)
+
+        def test_messages(mtype, expected):
+            messages = getattr(self.err, mtype)
+
+            if isinstance(expected, (list, tuple)):
+                assert all(find_message(messages, props)
+                           for props in expected)
+            elif expected:
+                assert messages, "Expected %s." % mtype
+            elif expected is not None:
```

Comment:
It replaces the docstring for `assert_failed`, but like I said, I added it just after I pushed. It will be in my next update later today.